src/MLStyle.jl

```diff
@@ -94,13 +94,108 @@ p                end
     end
 end
 
+#=
+    help functions for `@when`
+=#
+
+@active IsMacro{s::String}(x) begin
+    @match x begin
+        Expr(:macrocall, &(Symbol("@", s)), ::LineNumberNode) => true
+        Expr(:macrocall, &(Symbol("@", s)), ::LineNumberNode, arg) => true
+        _ => false
+    end
+end
+
+@active MacroSplit{s::String}(x) begin
+    @match x begin
+        Expr(:macrocall, &(Symbol("@", s)), ::LineNumberNode, arg) => arg
+        _ => nothing
+    end
+end
+
+function RecursionMatch(block, bds=[], rets=[], default=nothing)
+    @match block begin
+        # empty block
+        Expr(:block) => (bds, rets, default)
+        
+        # skip `LineNumberNode`
+        Expr(:block, 
+            ::LineNumberNode,
+            remain...
+        ) => 
+            RecursionMatch(Expr(:block, remain...), bds, rets, default)
+        
+        # match `@when`s at middle
+        Expr(:block, 
+            MacroSplit{"when"}(bd),
+            ::LineNumberNode,
+            ret,
+            remain...
+        ) && Do(push!(bds, bd)) && Do(push!(rets, ret)) => 
+            RecursionMatch(Expr(:block, remain...), bds, rets, default)
+        
+        # match `@when`s at end
+        Expr(:block, 
+            MacroSplit{"when"}(bd),
+            ::LineNumberNode,
+            ret
+        ) && Do(push!(bds, bd)) && Do(push!(rets, ret)) => 
+            (bds, rets, default)
+        
+        # match `@otherwise` at end
+        Expr(:block, 
+            IsMacro{"otherwise"}(),
+            ::LineNumberNode,
+            default
+        ) => (bds, rets, default)
+
+        # error handle
+        
+        _ => @syntax_err "Expect a :block."
+    end
+end
+
+macro otherwise() end
+
+
+
 """
 Code generation for `@when`.
 You should pass an `Expr(:let, ...)` as the first argument.
 """
 function gen_when(let_expr, source :: LineNumberNode, mod :: Module)
     @match let_expr begin
-        Expr(:let, Expr(:block, bindings...) ||  a && Do(bindings = [a]), ret) =>
+        # `@otherwise`
+        Expr(:let, 
+            Expr(:block, bindings...) || a && Do(bindings = [a]),
+            Expr(:block, 
+                ::LineNumberNode,
```

Comment:
Also, we shouldn't expect there is a LineNumberNode, for even if there's no LineNumberNode we can still compile the block expression.